lib/vrl/parser/src/lex.rs

```diff
@@ -436,17 +440,84 @@ impl<'input> Token<&'input str> {
 }
 
 #[derive(Clone, PartialEq, Eq, Debug, Hash)]
-pub enum StringLiteral<S> {
-    Escaped(S),
-    Raw(S),
-}
+pub struct StringLiteral<S>(pub S);
+
+#[derive(Clone, PartialEq, Eq, Debug, Hash)]
+pub struct RawStringLiteral<S>(pub S);
 
 impl StringLiteral<&str> {
-    pub fn unescape(&self) -> String {
-        match self {
-            StringLiteral::Escaped(s) => unescape_string_literal(s),
-            StringLiteral::Raw(s) => s.to_string(),
+    /// Takes the string and splits it into segments of literals and templates.
+    /// A templated section is delimited by `{{..}}`. ``{{` can be escaped using
+    /// `/{{.../}}`.
+    pub fn template(&self, span: Span) -> TemplateString {
+        let mut segments = Vec::new();
+
+        let chars = self.0.chars().collect::<Vec<_>>();
+        let mut template = false;
+        let mut current = String::new();
+
+        let mut pos = 0;
+        while pos < chars.len() {
+            match chars[pos] {
+                '}' if template && chars.get(pos + 1) == Some(&'}') => {
+                    // Handle closing template `}}`.
+                    if !current.is_empty() {
+                        let seg = std::mem::take(&mut current);
+                        segments.push(StringSegment::Template(
+                            seg.trim().to_string(),
+                            Span::new(span.start() + pos - seg.len() - 1, span.start() + pos + 3),
+                        ));
+                    }
+                    template = false;
+                    pos += 2;
+                }
+                '\\' if !template
+                    && chars.get(pos + 1) == Some(&'{')
+                    && chars.get(pos + 2) == Some(&'{') =>
+                {
+                    // Handle open escape `/{{`.
+                    current.push_str(r#"{{"#);
+                    pos += 3;
+                }
+                '\\' if !template
+                    && chars.get(pos + 1) == Some(&'}')
+                    && chars.get(pos + 2) == Some(&'}') =>
+                {
+                    // Handle close escape
+                    current.push_str(r#"}}"#);
+                    pos += 3;
+                }
+                '{' if !template && chars.get(pos + 1) == Some(&'{') => {
+                    // Handle start of template.
+                    if !current.is_empty() {
+                        let seg = std::mem::take(&mut current);
+                        segments.push(StringSegment::Literal(unescape_string_literal(&seg)));
+                    }
+                    template = true;
+                    pos += 2;
+                }
+                chr => {
+                    current.push(chr);
+                    pos += 1;
+                }
+            }
         }
+
+        if !template && !current.is_empty() {
```

Comment:
Doesn't this mean an empty string is always marked as a template string?

Reply:
If we have an empty string we have no segments. This is then handled [here](https://github.com/vectordotdev/vector/blob/stephen/vrl_interpolation/lib/vrl/parser/src/template_string.rs#L34-L39) to create an empty literal.

Reply:
That's true, but it does strike me as odd that we'd mark an empty string as a templated string in this piece of the code, regardless if we solve that in a separate place. It seems to me we want to ensure it's correctly marked here, to avoid any potential confusion if someone else uses this code without rewriting it in the future.

Reply:
All strings are marked as template strings at this stage. It's only when rewriting that it gets rewritten to either a literal string or a string concatenation operation.

Reply:
Ah yes, you are correct 👍

On first thought, that is somewhat of an odd approach, but given that it works, and rewriting it would probably require some rethinking here, I think this is fine to land as-is.